### PR TITLE
Token Lists by Network: the Comeback

### PR DESCRIPTION
**Pull request by W3stside**
## I know this PR is massive, bear with me

This is a rework of #107 and is different:
1. removal of custom `NetworkListUpdater` inside `custom/state/lists/updater/NetworkListUpdater` - why change from this? Because `NetworkListUpdater` introduces custom logic that is separated from the base app and isn't well tested considering the intricate logic of the token lists. Was very hard to debug if something went wrong
2. Mods `state/lists/reducer` to allow mapping `chainId` to a slice of token list state, this just adds a top level 
```ts
// Before:
type OldListsState = {
  byUrl: ...,
  lastInitialised: ...
  ... // etc
  // not accessible by chain
}

// Now:
type ListsState = [key in ChainId]: ListsState
// each chain maps to it's own lists state
```
3. Mods `state/lists/actions` to give action params `{ chainId: ChainId }` prop to designate which chain to save to and work with new reducer state
4. Mods `state/lists/updater` to create a new `useEffect` which calls `updateVersion({ chainId })` when chainId changes, initialising the lists for that new chain. This was effectively what the OLD PR was doing #107 inside the `NetworkListUpdater`
5. Mods various files dependent on the OLD `@src/state/lists/actions` and passes in the new actions with `ChainId`

### Tests
1. `custom/state/lists/reducer/reducerMod.test.ts` is updated to new state shape and passes
2. can also manually test by changing settings in token list under `Manage` when selecting a token, changing chain, going back to prev. chain and seeing persistent changes
3. same as above but leave tab and come back

### CODE REVIEW

File: src/custom/components/SearchModal/ManageLists/index.tsx [hunk 1]
```diff
@@ -4,12 +4,26 @@ import { DEFAULT_NETWORK_FOR_LISTS, UNSUPPORTED_LIST_URLS } from 'constants/list
...
+}
 
 export const ManageLists = (props: {
   setModalView: (view: CurrencyModalView) => void
   setImportList: (list: TokenList) => void
   setListUrl: (url: string) => void
 }) => {
   const { chainId = DEFAULT_NETWORK_FOR_LISTS } = useActiveWeb3React()
-  return <ManageListsMod {...props} unsupportedListUrls={UNSUPPORTED_LIST_URLS[chainId]} />
+  const listRowProps = {
```
Comment (W3stside) [inline src/custom/components/SearchModal/ManageLists/index.tsx:22]: to minimise changes necessary inside the old comp, we build the `lists/actions` functions here to retain the same shape as previously but pass in `chainId` as sort of a wrapped function

File: src/custom/state/lists/actions/actionsMod.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { ActionCreatorWithPayload, createAction } from '@reduxjs/toolkit'
+import { ChainId } from '@uniswap/sdk'
+import { TokenList, Version } from '@uniswap/token-lists'
+
+export interface WithChainId {
```
Comment (W3stside) [inline src/custom/state/lists/actions/actionsMod.ts:5]: the biggest MOD here is changing the types to take `chainId`

File: src/custom/state/lists/hooks/hooksMod.ts [hunk 3]
```diff
@@ -1,6 +1,119 @@
...
+import { DEFAULT_NETWORK_FOR_LISTS, UNSUPPORTED_LIST_URLS } from 'constants/lists'
+import { TokenList } from '@uniswap/token-lists'
+import DEFAULT_TOKEN_LIST from '@uniswap/default-token-list'
+import { TokenAddressMap, listToTokenMap, combineMaps, EMPTY_LIST } from '@src/state/lists/hooks'
+import sortByListPriority from 'utils/listSort'
+import UNSUPPORTED_TOKEN_LIST from 'constants/tokenLists/uniswap-v2-unsupported.tokenlist.json'
 
-export function useFullLists() {
-  return useSelector<AppState, AppState['lists']>(state => state.lists)
+export * from '@src/state/lists/hooks'
```
Comment (W3stside) [inline src/custom/state/lists/hooks/hooksMod.ts:13]: MOD: the previous `hooks` in the default files were using the scoped functions. Modding anything required me to copy the functions used in scope here. 

The rest of the functions are updated to use `chainId` via `useActiveWeb3React` hook

File: src/custom/state/lists/hooks/hooksMod.ts [hunk 4]
```diff
@@ -1,6 +1,119 @@
...
+export function useAllLists(): {
+  readonly [url: string]: {
+    readonly current: TokenList | null
+    readonly pendingUpdate: TokenList | null
+    readonly loadingRequestId: string | null
+    readonly error: string | null
+  }
+} {
+  // MOD: adds { chainId } support to the hooks
+  const { chainId = DEFAULT_NETWORK_FOR_LISTS } = useActiveWeb3React()
```
Comment (W3stside) [inline src/custom/state/lists/hooks/hooksMod.ts:27]: the main MOD is this line, and the comment below (repeats a few times in different functions here)

File: src/custom/state/lists/hooks/hooksMod.ts [hunk 5]
```diff
@@ -1,6 +1,119 @@
...
+  // return useSelector<AppState, AppState['lists']['byUrl']>(state => state.lists.byUrl)
+  return useSelector<AppState, AppState['lists'][ChainId]['byUrl']>(state => state.lists[chainId].byUrl)
```
Comment (W3stside) [inline src/custom/state/lists/hooks/hooksMod.ts:29]: another main MOD: adding the `[ChainId]` key access to the `useSelector` type

File: src/custom/state/lists/reducer/reducerMod.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+import { createReducer } from '@reduxjs/toolkit'
+import { getVersionUpgrade, VersionUpgrade } from '@uniswap/token-lists'
+import { TokenList } from '@uniswap/token-lists/dist/types'
+// import { DEFAULT_LIST_OF_LISTS } from '@src/constants/lists'
+import { updateVersion } from 'state/global/actions'
+import { acceptListUpdate, addList, fetchTokenList, removeList, enableList, disableList } from 'state/lists/actions'
+import { ChainId } from '@uniswap/sdk'
+
+// Mod: change state shape - adds network map
+export type ListsStateByNetwork = {
```
Comment (W3stside) [inline src/custom/state/lists/reducer/reducerMod.ts:16]: as noted in description: key access to state using `ChainId`

File: src/custom/state/lists/reducer/reducerMod.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  byUrl: {
+    ...DEFAULT_LIST_OF_LISTS_BY_NETWORK[chainId].reduce<Mutable<ListsState['byUrl']>>((memo, listUrl) => {
+      memo[listUrl] = NEW_LIST_STATE
+      return memo
+    }, {})
+  },
+  activeListUrls: DEFAULT_ACTIVE_LIST_URLS_BY_NETWORK[chainId]
+})
+
+const ChainIdList = Object.values(ChainId)
```
Comment (W3stside) [inline src/custom/state/lists/reducer/reducerMod.ts:59]: create a dynamic `list` of our `ChainId` using the enum (sorta hacky)

File: src/custom/state/lists/reducer/reducerMod.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+// MOD: change the intiialState shape
+// we make an object with each chainId pulled from ChainId enum
+// into a list and reduced into a map
+const initialState: ListsStateByNetwork = {
+  // cut in half as enums are always represented as key/value and then inverted
+  // https://stackoverflow.com/a/51536142
+  ...(ChainIdList.slice(ChainIdList.length / 2) as ChainId[]).reduce((memo, chainId) => {
```
Comment (W3stside) [inline src/custom/state/lists/reducer/reducerMod.ts:66]: bit hacky but:
```ts
enum Net {
  "MAINNET": 1,
  "RINKEBY": 4,
}

const ChainIdList = Object.values(ChainId)
// ["MAINNET", "RINKEBY", 1, 4]

const list = ChainIdList.slice(ChainIdList.length / 2)
// [1, 4]
```

when `enum` changes, this auto adapts

File: src/custom/state/lists/reducer/reducerMod.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    return memo
+  }, {} as ListsStateByNetwork)
+}
+
+export default createReducer(initialState, builder =>
+  builder
+    .addCase(
+      fetchTokenList.pending,
+      (baseState, { payload: { chainId = DEFAULT_NETWORK_FOR_LISTS, requestId, url } }) => {
+        const state = baseState[chainId]
```
Comment (W3stside) [inline src/custom/state/lists/reducer/reducerMod.ts:79]: only changes here are adapting to the new state shape, nothing else

File: src/custom/state/lists/reducer/reducerMod.ts [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    }
+    return memo
+  }, {} as ListsStateByNetwork)
+}
+
+export default createReducer(initialState, builder =>
+  builder
+    .addCase(
+      fetchTokenList.pending,
+      (baseState, { payload: { chainId = DEFAULT_NETWORK_FOR_LISTS, requestId, url } }) => {
```
Comment (W3stside) [inline src/custom/state/lists/reducer/reducerMod.ts:78]: defaults to `constant/index` `DEFAULT_NETWORK_FOR_LISTS` which here = `ChainId.MAINNET`

File: src/custom/state/lists/updater/updaterMod.ts [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            dispatch(acceptListUpdate({ chainId, url: listUrl }))
+        }
+      }
+    })
+    // MOD: deps
+    // }, [dispatch, lists, activeListUrls])
+  }, [dispatch, lists, activeListUrls, chainId])
+
+  // automatically initialise lists if chainId changes
+  useEffect(() => {
```
Comment (W3stside) [inline src/custom/state/lists/updater/updaterMod.ts:89]: Only update here that isn't adapting to new state shape

this calls list initialisation logic on chainId change

File: src/components/SearchModal/ImportList.tsx [hunk 12]
```diff
@@ -17,7 +17,7 @@ import { TokenList } from '@uniswap/token-lists'
 import { useDispatch } from 'react-redux'
 import { AppDispatch } from 'state'
 import { useFetchListCallback } from 'hooks/useFetchListCallback'
-import { removeList, enableList } from 'state/lists/actions'
+import { removeList, enableList } from '@src/state/lists/actions'
```
Comment (anxolin) [inline src/components/SearchModal/ImportList.tsx:20]: why do you use '@src/state/lists/actions' ?

Once you override a module, like in this case, I would use our custom

File: src/custom/components/Popups/ListUpdatePopup.tsx [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+  popKey: string
+  listUrl: string
+  oldList: TokenList
+  newList: TokenList
+  auto: boolean
+  // MOD: accept custom callback
+  acceptListUpdate: ListRowProps['acceptListUpdate']
+}
+
+export default function ListUpdatePopup(props: Omit<ListUpdatePopupProps, 'acceptListUpdate'>) {
```
Comment (anxolin) [inline src/custom/components/Popups/ListUpdatePopup.tsx:19]: perfect override 👌

File: src/custom/components/Popups/ListUpdatePopup.tsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
```
Comment (anxolin) [inline src/custom/components/Popups/ListUpdatePopup.tsx:1]: for some files, when we override we create the dir, the index file and the mod file. 
Here we didn't do the directory. Im fine with this too, is ok to decide on a per case basis or should we create the dir to really group the overrided component and it's mod

File: src/custom/state/lists/hooks/hooksMod.ts [hunk 15]
```diff
@@ -1,6 +1,119 @@
...
+import DEFAULT_TOKEN_LIST from '@uniswap/default-token-list'
+import { TokenAddressMap, listToTokenMap, combineMaps, EMPTY_LIST } from '@src/state/lists/hooks'
+import sortByListPriority from 'utils/listSort'
+import UNSUPPORTED_TOKEN_LIST from 'constants/tokenLists/uniswap-v2-unsupported.tokenlist.json'
 
-export function useFullLists() {
-  return useSelector<AppState, AppState['lists']>(state => state.lists)
+export * from '@src/state/lists/hooks'
+
+// MOD
```
Comment (anxolin) [inline src/custom/state/lists/hooks/hooksMod.ts:15]: You deleted here lots of lines of this mod:
image

You can leave the lines, and comment them if you don't need them

**Review by anxolin — COMMENTED** [review]
You are becoming an expert on hacking the app. This PR feels that the Mods have better quality (they are more precise on what they change)

I didn't finish the review, but I felt like giving some feedback already, I've reviewed 18 files already... Until `hooksMod` it was all looking good to me

image

File: src/components/SearchModal/ImportList.tsx (hunk 12 above)
Comment (W3stside) [reply src/components/SearchModal/ImportList.tsx:20]: in this case this is an escape hatch

Comment (W3stside) [reply src/components/SearchModal/ImportList.tsx:20]: i need them to use the defaults

File: src/custom/components/Popups/ListUpdatePopup.tsx (hunk 13 above)
Comment (W3stside) [reply src/custom/components/Popups/ListUpdatePopup.tsx:19]: :+1:

File: src/custom/components/Popups/ListUpdatePopup.tsx (hunk 14 above)
Comment (W3stside) [reply src/custom/components/Popups/ListUpdatePopup.tsx:1]: :+1: my thoughts here as well

File: src/custom/state/lists/hooks/hooksMod.ts (hunk 15 above)
Comment (W3stside) [reply src/custom/state/lists/hooks/hooksMod.ts:15]: is it optimal to do this when I'm not deleting them but simply importing them from the base file? regardless i copied them back in and left them commented out

File: src/components/SearchModal/ImportList.tsx (hunk 12 above)
Comment (anxolin) [reply src/components/SearchModal/ImportList.tsx:20]: aha, ok, then it make sense

File: src/custom/state/lists/hooks/hooksMod.ts (hunk 15 above)
Comment (anxolin) [reply src/custom/state/lists/hooks/hooksMod.ts:15]: The rule we don't want to breake is:
`The mod always shoud compare line by line.` so we don't delete

I guess you refer to:
```
export * from '@src/state/lists/hooks'
```

That line above should be used in the `index.ts`, and then, the things you don't use in the Mod, should be there but commented out. 

Also, 
image

Comment (W3stside) [reply src/custom/state/lists/hooks/hooksMod.ts:13]: ok the changes i made address this

Comment (W3stside) [reply src/custom/state/lists/hooks/hooksMod.ts:13]: https://github.com/gnosis/gp-swap-ui/pull/244/commits/4b83cbbb73353d9f672e159dd3225fa2bd2e1589

File: src/custom/state/lists/reducer/reducerMod.ts (hunk 8 above)
Comment (anxolin) [reply src/custom/state/lists/reducer/reducerMod.ts]: I just moved it to an util 4d8763b9aea29efcb24ce8eb6fe40855bc8f6520

I don't like the code, but I can't suggest sth better now, so I hide it inside an util

**Review by anxolin — APPROVED** [review]
😅 it was a hard review. I'm sure it drained also a lot of energy doing it.

**Pull request merged**